File: sampleapps/swift/StrongKeyFIDODemo/Shared/FIDOService/FidoService.swift
/**
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License, as published by the Free Software Foundation and
 * available at http://www.fsf.org/licensing/licenses/lgpl.html,
 * version 2.1.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * Copyright (c) 2001-2021 StrongAuth, Inc. (DBA StrongKey)
 *
 * **********************************************
 *
 * 888b    888          888
 * 8888b   888          888
 * 88888b  888          888
 * 888Y88b 888  .d88b.  888888  .d88b.  .d8888b
 * 888 Y88b888 d88""88b 888    d8P  Y8b 88K
 * 888  Y88888 888  888 888    88888888 "Y8888b.
 * 888   Y8888 Y88..88P Y88b.  Y8b.          X88
 * 888    Y888  "Y88P"   "Y888  "Y8888   88888P'
 *
 * **********************************************
 *
 * FidoService handles all the communication with the SKFS Servlet
 *
 * See SKFS FIDO Server https://github.com/StrongKey/fido2 for details.
 *
 */

import SwiftUI

/// LoginService handles all the communication with the SKFS Servlet
class FidoService {
    
    /// PreRegister is used to fetch the challenge from SKFS for the user
    /// - Parameters:
    ///   - username: Unique User Name set by the account user
    ///   - displayName: Display name can be the Full Name of the user or any other name of their choice
    ///   - completion: Passes the Data, Response and Error to the View where the function is called
    func preRegister(username: String, displayName: String, completion: @escaping (PreRegResponseModel?, URLResponse?, Error?) -> Void) {
        let preRegisterEndpoint = APICases.preRegister(username: username, displayName: displayName)
        request(endpoint: preRegisterEndpoint) { data, response, error in
            
            guard let data = data else {
                print(String(describing: error))
                return
            }
            
            guard let httpResponse = response as? HTTPURLResponse,
                  (200...299).contains(httpResponse.statusCode) else {
                      let errorResponseModel = try! JSONDecoder().decode(ErrorResponseModel.self, from: data)
                      print(errorResponseModel.message)
                      return completion(nil, response, error)
                  }
            
            let responseJSON = try! JSONSerialization.jsonObject(with: data, options: []) as! [String: Any]
            let responseDataString = responseJSON["Response"] as! String
            let responseData = Data(responseDataString.utf8)
            
            let dataFromPreRegModel: PreRegResponseModel = try! JSONDecoder().decode(PreRegResponseModel.self, from: responseData)
            
            // MARK: Change the completion to NOT send the preRegResponseModel data to the view controller
            completion(dataFromPreRegModel, response, error)
        }
    }
    
    /// Register endpoint verifies the attestation from the Platform Authenticator
    /// - Parameters:
    ///   - username: username input from the user
    ///   - attestationObject: attestationObject from the Sign Up Request
    ///   - clientDataJSON: clientDataJSON has the information about the authenticator
    ///   - id: Credential ID from the Platform Authenticator
    ///   - completion: completion handler provides the response from the SKFS
    func register(username: String, attestationObject: String, clientDataJSON: String, id: String, completion: @escaping (Data?, URLResponse?, Error?) -> Void) {
        
        let registerEndpoint = APICases.register(username: username, attestationObject: attestationObject, clientDataJSON: clientDataJSON, id: id)
        
        request(endpoint: registerEndpoint) { data, response, error in
            completion(data, response, error)
        }
        
    }
    
    /// Register Existing is used to verify attestation for adding additional Keys for the user to login on other devices or other security reasons
    /// - Parameters:
    ///   - username: username of the currently logged in user
    ///   - attestationObject: attestationObject from the authenticator
    ///   - clientDataJSON: clientDataJSON from the authenticator
    ///   - id: Credential ID
    ///   - jwt: Stored JWT of the currently logged in user from AppStorage
    ///   - completion: completion handler provides the response from the SKFS
    func registerExisting(username: String, attestationObject: String, clientDataJSON: String, id: String, jwt: String, completion: @escaping (Data?, URLResponse?, Error?) -> Void) {
        
        let registerExistingEndpoint = APICases.registerExisting(username: username, attestationObject: attestationObject, clientDataJSON: clientDataJSON, id: id, jwt: jwt)
        
        request(endpoint: registerExistingEndpoint) { data, response, error in
            completion(data, response, error)
        }
        
    }
    
    /// Pre Authenticate gets the challenge for the specified user from the server
    /// - Parameters:
    ///   - username: username of the user
    ///   - completion: completion returns PreAuthResponseModel which has the challenge for attestation
    func preauthenticate(username: String, completion: @escaping (PreAuthResponseModel?, URLResponse?, Error?) -> Void) {
        
        let preAuthenticateEndpoint = APICases.preauthenticate(username: username)
        
        request(endpoint: preAuthenticateEndpoint) { data, response, error in
            guard let data = data else {
                print(String(describing: error))
                return
            }
            
            guard let httpResponse = response as? HTTPURLResponse,
                  (200...299).contains(httpResponse.statusCode) else {
                      let errorResponseModel = try! JSONDecoder().decode(ErrorResponseModel.self, from: data)
                      print(errorResponseModel.message)
                      return completion(nil, response, error)
                  }
            
            //            print(data)
            
            let responseJSON = try! JSONSerialization.jsonObject(with: data, options: []) as! [String: Any]
            let responseDataString = responseJSON["Response"] as! String
            let responseData = Data(responseDataString.utf8)
            
            //            print(responseDataString)
            
            let dataFromPreAuthModel: PreAuthResponseModel = try! JSONDecoder().decode(PreAuthResponseModel.self, from: responseData)
            
            completion(dataFromPreAuthModel, response, error)
            
        }
        
    }
    
    /// Authenticate the user by verifying the signature from the server
    /// - Parameters:
    ///   - username: username of the currently logged in user
    ///   - authenticatorData: attestationObject from the authenticator
    ///   - signature: signature from the authenticator
    ///   - userHandle: Empty string in U2F
    ///   - clientDataJSON: clientDataJSON from the authenticator
    ///   - id: Credential ID on device
    ///   - completion: completion handler returns AuthResponseModel which has the JWT in it which needs to be stored in App Storage for future use
    func authenticate(username: String, authenticatorData: String, signature: String, userHandle: String, clientDataJSON: String, id: String, completion: @escaping (AuthResponseModel?, URLResponse?, Error?) -> Void) {
        
        let authenticateEndpoint = APICases.authenticate(username: username, authenticatorData: authenticatorData, signature: signature, userHandle: userHandle, clientDataJSON: clientDataJSON, id: id)
        
        request(endpoint: authenticateEndpoint) { data, response, error in
            
            guard let data = data else {
                print(String(describing: error))
                return
            }
            
            guard let httpResponse = response as? HTTPURLResponse,
                  (200...299).contains(httpResponse.statusCode) else {
                      let errorResponseModel = try! JSONDecoder().decode(ErrorResponseModel.self, from: data)
                      print(errorResponseModel.message)
                      return completion(nil, response, error)
                  }
            
            let dataFromAuthModel: AuthResponseModel = try! JSONDecoder().decode(AuthResponseModel.self, from: data)
            
            completion(dataFromAuthModel, response, error)
        }
        
    }
    
    /// Get details of all the keys registered to the account
    /// - Parameters:
    ///   - username: username of the currently logged in user
    ///   - jwt: JWT for the user saved during Authentication
    ///   - completion: completion handler returns a KeyInfoModel which has all keys registered to an account and their details
    func getKeysInfo(username: String, jwt: String, completion: @escaping (KeyInfoModel?, URLResponse?, Error?) -> Void) {
        let getkeysinfoEndpoint = APICases.getKeysInfo(username: username, jwt: jwt)
        
        request(endpoint: getkeysinfoEndpoint) { data, response, error in
            guard let data = data else {
                print(String(describing: error))
                return
            }
            
            guard let httpResponse = response as? HTTPURLResponse,
                  (200...299).contains(httpResponse.statusCode) else {
                      let errorResponseModel = try! JSONDecoder().decode(ErrorResponseModel.self, from: data)
                      print(errorResponseModel.message)
                      return completion(nil, response, error)
                  }
            
            let responseJSON = try! JSONSerialization.jsonObject(with: data, options: []) as! [String: Any]
            
            let responseDataString = responseJSON["Response"] as! String
            let responseData = Data(responseDataString.utf8)
            
            let dataFromPreAuthModel: KeyInfoModel = try! JSONDecoder().decode(KeyInfoModel.self, from: responseData)
            
            completion(dataFromPreAuthModel, response, error)
            
        }
    }
    
    
    
    /// Get challenge for currently logged in user using JWT verification to add additional FIDO Keys
    /// - Parameters:
    ///   - username: username of currently logged in user
    ///   - displayName: displayName of the currently logged in user
    ///   - jwt: Stored JWT of the currently logged in user from AppStorage
    ///   - completion: Get the challenge from completion handler to send in registerExisitng endpoint
    func preRegisterExisting(username: String, displayName: String, jwt: String, completion: @escaping (PreRegResponseModel?, URLResponse?, Error?) -> Void) {
        let preRegisterExistingEndpoint = APICases.preRegisterExisting(username: username, displayName: displayName, jwt: jwt)
        
        request(endpoint: preRegisterExistingEndpoint) { data, response, error in
            guard let data = data else {
                print(String(describing: error))
                return
            }
            
            guard let httpResponse = response as? HTTPURLResponse,
                  (200...299).contains(httpResponse.statusCode) else {
                      let responseJSON = try! JSONSerialization.jsonObject(with: data, options: []) as! [String: Any]
                      print(responseJSON)
                      return
                  }
            
            let responseJSON = try! JSONSerialization.jsonObject(with: data, options: []) as! [String: Any]
            let responseDataString = responseJSON["Response"] as! String
            let responseData = Data(responseDataString.utf8)
            
            let dataFromPreRegModel: PreRegResponseModel = try! JSONDecoder().decode(PreRegResponseModel.self, from: responseData)
            
            completion(dataFromPreRegModel, response, error)
            
        }
    }
    
    /// Remove the selected key from a user account
    /// - Parameters:
    ///   - username: username of currently logged in user
    ///   - keyIds: randomID of the key
    ///   - jwt: Stored JWT of the currently logged in user from AppStorage
    ///   - completion: completion handler returns the response from SKFS Servlet
    func removeKeys(username: String, keyIds: [String], jwt: String, completion: @escaping (Data?, URLResponse?, Error?) -> Void ) {
        
        let removeKeysEnpoint = APICases.removeKeys(username: username, keyIDs: keyIds, jwt: jwt)
        
        request(endpoint: removeKeysEnpoint) { data, response, error in
            guard let data = data else {
                print(String(describing: error))
                return
            }
            
            guard let httpResponse = response as? HTTPURLResponse, (200...299).contains(httpResponse.statusCode) else {
                let responseJSON = try! JSONSerialization.jsonObject(with: data, options: []) as! [String: Any]
                print(responseJSON)
                return
            }
            
            completion(data, response, error)
            
            
        }
        
    }
    
    
    /// Deletes the user account and all its registered keys
    /// - Parameters:
    ///   - username: username of currently logged in user
    ///   - jwt: Stored JWT of the currently logged in user from AppStorage
    ///   - completion: completion handler returns the response from SKFS Servlet
    func deleteAccount(username: String, jwt: String, completion: @escaping (Data?, URLResponse?, Error?) -> Void) {
        
        let deleteAccountEndpoint = APICases.deleteAccount(username: username, jwt: jwt)
        
        request(endpoint: deleteAccountEndpoint) { data, response, error in
            
            guard let data = data else {
                print(String(describing: error))
                return
            }
            
            guard let httpResponse = response as? HTTPURLResponse, (200...299).contains(httpResponse.statusCode) else {
                let responseJSON = try! JSONSerialization.jsonObject(with: data, options: []) as! [String: Any]
                print(responseJSON)
                return
            }
           
            completion(data, response, error)
            
        }
    }
    
    
    
    
    
    /// This is used to create endpoint request and make the actual URLSession Request and send the data back in the completion handler
    /// - Parameters:
    ///   - endpoint: Enum to which Endpoint to make the URLSession request
    ///   - completion: completion handler sends back data, response and error back to the caller
    private func request(endpoint: Endpoint, completion: @escaping (Data?, URLResponse?, Error?) -> Void) {
        
        // URL
        let url = URL(string: endpoint.url)!
        var urlRequest = URLRequest(url: url)
        
        // HTTP Method
        urlRequest.httpMethod = endpoint.httpMethod
        
        // Header fields
        endpoint.headers?.forEach({ header in
            urlRequest.setValue(header.value as? String, forHTTPHeaderField: header.key)
        })
        
        // HTTP Body
        //        urlRequest.setValue("Application/json", forHTTPHeaderField: "Content-Type")
        guard let httpBody = try? JSONSerialization.data(withJSONObject: endpoint.body!, options: []) else {
            return
        }
        
        urlRequest.httpBody = httpBody
        
        let task = URLSession.shared.dataTask(with: urlRequest) { data, response, error in
            completion(data, response, error)
        }
        
        task.resume()
    }
    
    // MARK: Uncomment this when running the FIDO2 server locally to test internal endpoints
    //    func urlSession(_ session: URLSession, didReceive challenge: URLAuthenticationChallenge, completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void) {
    //        if challenge.protectionSpace.serverTrust == nil {
    //            completionHandler(.useCredential, nil)
    //        } else {
    //            let trust: SecTrust = challenge.protectionSpace.serverTrust!
    //            let credential = URLCredential(trust: trust)
    //            completionHandler(.useCredential, credential)
    //        }
    //    }
    
}

File: sampleapps/java/sacl/mobile/android/StrongKeyAndroidClientLibrary/sacl/src/main/java/com/strongkey/sacl/interfaces/FidoService.java
/**
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License, as published by the Free Software Foundation and
 * available at http://www.fsf.org/licensing/licenses/lgpl.html,
 * version 2.1.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * Copyright (c) 2001-2021 StrongAuth, Inc. (DBA StrongKey)
 *
 * **********************************************
 *
 *  888b    888          888
 *  8888b   888          888
 *  88888b  888          888
 *  888Y88b 888  .d88b.  888888  .d88b.  .d8888b
 *  888 Y88b888 d88""88b 888    d8P  Y8b 88K
 *  888  Y88888 888  888 888    88888888 "Y8888b.
 *  888   Y8888 Y88..88P Y88b.  Y8b.          X88
 *  888    Y888  "Y88P"   "Y888  "Y8888   88888P'
 *
 * **********************************************
 *
 * An interface that represents FIDO webservices available on the
 * StrongKey FIDO Server.
 *
 * It should be noted that a rich-client app (RCA) using this library
 * will NOT be calling the FIDO webservices directly. RCAs, typically,
 * interact with a business application and as such, must communicate
 * with that application for its FIDO registration, authentications and
 * transaction confirmations.
 *
 * StrongKey's sample open-source implementation of a Mobile Device
 * Business Application (MDBA) is used as an example to demonstrate the
 * use of this library. While the MDBA serves no real purpose (other than
 * to demonstrate how to use this library for a variety of use-cases),
 * the MDBA can be leveraged to build other RCAs that perform useful
 * business functions while using this library for cryptographic services.
 *
 * The design goal of this library to simplify the complexity of various
 * cryptographic protocols through higher level application programming
 * interfaces (API) that serve business app developers; obviously, those
 * who want far more control over their app's cryptographic capability can
 * find their way to various standards sites/protocols for lower-level APIs.
 * Be warned: you should be prepared to get your hands really dirty.
 *
 * COMMERCIAL DECLARATION: While StrongKey has been a strong proponent of
 * the open-source movement since inception (2001), it must rely on some
 * revenue to be able to do the hard research and work that produces open-
 * source software such as our FIDO server and this library.
 *
 * As such, the MDBA is built on an assumption that security-conscious
 * relying parties (RP) such as banks, government agencies and companies in
 * the e-commerce, healthcare and other sectors dealing with sensitive data,
 * will choose to manage cryptographic capability that protects data within
 * their RCA on mobile devices, through policies dictated by a centralized
 * key management system (KMS). StrongKey does make such a KMS product
 * available to sustain its existence, and the MDBA is assumed to be
 * operating under the control of such a KMS. Consequently, there are
 * references to the KMS within the MDBA APIs.
 *
 * Developers who would like to use this library without such a KMS can
 * modify the source code in this library - operating under the control of
 * the open source license referenced above - and use/redistribute it under
 * the terms of that license.
 *
 * Thank you for your interest in downloading and using this library.
 */

package com.strongkey.sacl.interfaces;

import com.strongkey.sacl.roomdb.PreauthorizeChallenge;

import org.json.JSONObject;

import java.security.Signature;

public interface FidoService {
//    /**
//     * Gets a challenge to initiate a FIDO Registration process.
//     *
//     * @param did integer value representing the cryptographic domain ID to
//     * which this webservice is directed (see https://strongkey.com/resources
//     * for an explanation of StrongKey cryptographic domains)
//     * @param uid long value representing the User ID within the MDBA that is
//     * making this service request. Since the MDBA manages the registration
//     * process and relays the service request to the FIDO server, it knows what
//     * to send to the FIDO server
//     * @param devid long value representing the device ID of this mobile device.
//     * This is known to the MDBA and is part of the on-boarding process of the
//     * mobile device when it is enrolled in the MDKMS
//     * @param rdid long value representing the registered device ID of this
//     * mobile device within the MDKMS. This will determine the KMS policy that
//     * will guide this library's actions in concert with the RCA
//     * @return PreregisterChallenge A Java object that represents the challenge,
//     * along with metadata about the challenge and the transaction. Embedded
//     * within this Java object, are details of the challenge provided by
//     * StrongKey's preregister() webservice on the FIDO server
//     */
//    PreregisterChallenge getFidoRegistrationChallenge(int did, long uid, long devid, long rdid);

    /**
     * Registers a FIDO public key with the StrongKey FIDO server as part of
     * the FIDO Registration process.
     *
     * @param did integer value representing the cryptographic domain ID to
     * which this webservice is directed (see https://strongkey.com/resources
     * for an explanation of StrongKey cryptographic domains)
     * @param uid long value representing the User ID within the MDBA that is
     * making this service request. Since the MDBA manages the registration
     * process and relays the service request to the FIDO server, it knows what
     * to send to the FIDO server
     * @param devid long value representing the device ID of this mobile device.
     * This is known to the MDBA and is part of the on-boarding process of the
     * mobile device when it is enrolled in the MDKMS
     * @param rdid long value representing the registered device ID of this
     * mobile device within the MDKMS. This will determine the KMS policy that
     * will guide this library's actions in concert with the RCA
     * https://www.w3.org/TR/2019/REC-webauthn-1-20190304/#android-key-attestation
     * @return Object A Java object that is either a PublicKeyCredential, which
     * represents the newly minted key-pair within AndroidKeystore, along with
     * metadata about the transaction, or a JSONError indicating a problem.
     * The private key of the key-pair is always stored within AndroidKeystore,
     * protected either by the Trusted Execution Environment (TEE) or a
     * hardware security module aka Secure Element (SE)
     */
    public Object registerAttestedDeviceFidoKey(int did, long uid, long devid, long rdid);

    /**
     * Registers a FIDO public key with the StrongKey FIDO server as part of
     * the FIDO Registration process.
     *
     * @param did integer value representing the cryptographic domain ID to
     * which this webservice is directed (see https://strongkey.com/resources
     * for an explanation of StrongKey cryptographic domains)
     * @param uid long value representing the User ID within the SFA eCommerce app
     * making this service request. Since the SFA manages the registration
     * process and relays the service request to the FIDO server, it knows what
     * to send to the FIDO server
     * @return Object A Java object that is either a PublicKeyCredential, which
     * represents the newly minted key-pair within AndroidKeystore, along with
     * metadata about the transaction, or a JSONError indicating a problem.
     * The private key of the key-pair is always stored within AndroidKeystore,
     * protected either by the Trusted Execution Environment (TEE) or a
     * hardware security module aka Secure Element (SE)
     */
    public Object registerFidoKey(int did, long uid);


//    /**
//     * Gets a challenge to initiate a FIDO Authentication process. It
//     * presupposes that the user has a FIDO key registered with the FIDO
//     * server for this to work.
//     *
//     * @param did integer value representing the cryptographic domain ID to
//     * which this webservice is directed (see https://strongkey.com/resources
//     * for an explanation of StrongKey cryptographic domains)
//     * @param uid long value representing the User ID within the MDBA that is
//     * making this service request.  Since the MDBA manages the authentication
//     * process and relays the service request to the FIDO server, it knows what
//     * to send to the FIDO server
//     * @param devid long value representing the device ID of this mobile device.
//     * This is known to the MDBA and is part of the on-boarding process of the
//     * mobile device when it is enrolled in the MDKMS
//     * @param rdid long value representing the registered device ID of this
//     * mobile device within the MDKMS. This will determine the KMS policy that
//     * will guide this library's actions in concert with the RCA
//     * @return PreauthenticateChallenge A Java object that represents the
//     * challenge, along with metadata about the challenge and the transaction.
//     * Embedded within this Java object, are details of the challenge provided
//     * by StrongKey's preauthenticate() webservice on the FIDO server
//     */
//    PreauthenticateChallenge getFidoAuthenticationChallenge(int did, long uid, long devid, long rdid);

    /**
     * Authenticates the user with a digital signature response from the
     * previously registered FIDO key with the StrongKey FIDO server.
     *
     * @param did integer value representing the cryptographic domain ID to
     * which this webservice is directed (see https://strongkey.com/resources
     * for an explanation of StrongKey cryptographic domains)
     * @param uid long value representing the User ID within the MDBA making
     * this service request. Since the MDBA manages the authentication process
     * and relays the service request to the FIDO server, it knows what to send
     * to the FIDO server
     * @param devid long value representing the device ID of this mobile device.
     * This is known to the MDBA and is part of the on-boarding process of the
     * mobile device when it is enrolled in the MDKMS
     * @param rdid long value representing the registered device ID of this
     * mobile device within the MDKMS. This will determine the KMS policy that
     * guides this library's actions in concert with the RCA
//     * @param payload JSONObject containing the payload response with the
//     * digital signature from the previously registered FIDO key of the user to
     * challenge posed by the FIDO server. As always, the transaction flows to
     * the FIDO server through the business application - in this case, the MDBA
     * @return Object A Java object that is either a AuthenticationSignature,
     * which represents the FIDO digital signature, along with metadata about
     * the transaction, or a JSONError indicating a problem.
     */
    public Object authenticateAttestedDeviceFidoKey(int did, long uid, long devid, long rdid);

    /**
     * Authenticates the user with a digital signature response from the
     * previously registered FIDO key with the StrongKey FIDO server.
     *
     * @param did integer value representing the cryptographic domain ID to
     * which this webservice is directed (see https://strongkey.com/resources
     * for an explanation of StrongKey cryptographic domains)
     * @param uid long value representing the User ID within the SFA making
     * this service request. Since the SFA manages the authentication process
     * and relays the service request to the FIDO server, it knows what to send
     * to the FIDO server
     * @return Object A Java object that is either an AuthenticationSignature,
     * which represents the FIDO digital signature, along with metadata about
     * the transaction, or a JSONError indicating a problem.
     */
    public Object authenticateFidoKey(int did, long uid);

//    /**
//     * Gets a challenge to initiate a FIDO Transaction Confirmation process.
//     * It presupposes that the user has a FIDO key registered with the FIDO
//     * server for this to work. This is the only pre-* methods that requires
//     * a payload parameter since it needs to pass on a few elements of the
//     * transaction the FIDO server uses to derive a challenge; this is what
//     * ties the transaction to the FIDO digital signature
//     *
//     * @param did integer value representing the cryptographic domain ID to
//     * which this webservice is directed (see https://strongkey.com/resources
//     * for an explanation of StrongKey cryptographic domains)
//     * @param uid long value representing the User ID within the MDBA that is
//     * making this service request. Since the MDBA manages the authorization
//     * process and relays the service request to the FIDO server, it knows what
//     * to send to the FIDO server
//     * @param devid long value representing the device ID of this mobile device.
//     * This is known to the MDBA and is part of the on-boarding process of the
//     * mobile device when it is enrolled in the MDKMS
//     * @param rdid long value representing the registered device ID of this
//     * mobile device within the MDKMS. This will determine the KMS policy that
//     * will guide this library's actions in concert with the RCA
//     * @param payload JSONObject value containing elements of the transaction
//     * for which the RP desires a transaction confirmation incorporating the
//     * the user's acceptance to the transaction with a digital signature from
//     * from the previously registered FIDO key
//     * @return PreauthorizeChallenge A Java object that represents the
//     * challenge, along with metadata about the challenge and the transaction.
//     * Embedded within this Java object, are details of the challenge provided
//     * by StrongKey's preauthorize() webservice on the FIDO server
//     */
//    PreauthorizeChallenge getFidoAuthorizationChallenge(int did, long uid, long devid, long rdid, JSONObject payload);

    /**
     * Gets a challenge to initiate a FIDO Transaction Confirmation process.
     * It presupposes that the user has a FIDO key registered with the FIDO
     * server for this to work. This is the only pre-* methods that requires
     * a payload parameter since it needs to pass on a few elements of the
     * transaction the FIDO server uses to derive a challenge; this is what
     * ties the transaction to the FIDO digital signature
     *
     * @param did integer value representing the cryptographic domain ID to
     * which this webservice is directed (see https://strongkey.com/resources
     * for an explanation of StrongKey cryptographic domains)
     * @param uid long value representing the User ID within the MDBA that is
     * making this service request. Since the MDBA manages the authorization
     * process and relays the service request to the FIDO server, it knows what
     * to send to the FIDO server
     * @param cart String with a Base64Url encoded JSONObject containing elements
     * of the transaction for which the RP desires a transaction confirmation. The
     * sample cart has a collection of products and the payment method selected by
     * the user
     * @return PreauthorizeChallenge object that represents the FIDO challenge
     * provided by StrongKey's preauthorize() webservice on the FIDO server or
     * a JSONObject with an error
     */
    Object getFidoAuthorizationChallenge(int did, long uid, String cart);

    /**
     * Authorizes a business transaction using the FIDO key. The RP may use
     * the transaction data and digital signature for a variety of use-cases:
     * compliance with the EU PSD2 for SCA, conform to the EMVCo 3DS2 protocol
     * or other private transaction protocol that leverages FIDO
     *
     * @param did integer value representing the cryptographic domain ID to
     * which this webservice is directed (see https://strongkey.com/resources
     * for an explanation of StrongKey cryptographic domains)
     * @param uid long value representing the User ID within the MDBA making
     * this service request. Since the MDBA manages the authentication process
     * and relays the service request to the FIDO server, it knows what to send
     * to the FIDO server
     * @param txid String containing a locally generated transaction ID. This
     * should be replaced with a TXID generated by the business application
     * @param txplayload String containing a JSONObject with products and payment
     * information
     * @param credentialId String containing unique key identifier in keystore
     * @param challenge String with the FIDO challenge to sign
     * @param signature Signature object initialized with the user's FIDO key
     * to generate the digital signature over the transaction
     *
     * @return Object A Java object, AuthorizationSignature, that represents the
     * digital signature, along with metadata about the challenge and the
     * transaction, or a JSONError indicating a problem.
     */
    public Object authorizeFidoTransaction(int did, long uid, String txid, String txplayload,
                                    String credentialId, String challenge, Signature signature);

}
